heartbeat/openstack-virtual-ip

```diff
@@ -0,0 +1,275 @@
+	fi
+
+	. $OCF_RESKEY_openrc
+
+	${HA_SBIN_DIR}/attrd_updater --query -n openstack_ports -N $(crm_node -n) > /dev/null 2>&1
+	if [ $? -ne 0 ] ; then
+		ocf_log warn "attr_updater failed to get openstack_ports attribute of node $OCF_RESOURCE_INSTANCE"
+		return $OCF_ERR_GENERIC
+	fi
+
+	return $OCF_SUCCESS
+}
+
+osvip_monitor() {
+	local result
+
+	node_port_id=$(osvip_port_id)
+
+	result=$($OCF_RESKEY_openstackcli port show \
+		--format value \
+		--column allowed_address_pairs \
+		${node_port_id})
+	if echo $result | grep -q $OCF_RESKEY_ip ; then
+		${HA_SBIN_DIR}/attrd_updater ${OCF_RESKEY_delay} -S status -n openstack_virtual_ip -v $OCF_RESKEY_ip
+
+		return $OCF_SUCCESS
+	fi
+
+	${HA_SBIN_DIR}/attrd_updater ${OCF_RESKEY_delay} -D -S state -n openstack_virtual_ip
+	ocf_log warn "$OCF_RESKEY_ip is not attached to any fixed address"
+	return $OCF_NOT_RUNNING
+}
+
+osvip_stop() {
+	node_port_id=$(osvip_port_id)
+
+	ocf_log info "Bringing down IP address $OCF_RESKEY_ip"
+
+	osvip_monitor
+	if [ $? = $OCF_NOT_RUNNING ]; then
+		ocf_log info "Address $OCF_RESKEY_ip already down"
+		return $OCF_SUCCESS
+	fi
+
+	mac_address=$($OCF_RESKEY_openstackcli port show \
+		--format value \
+		--column mac_address \
+		$node_port_id)
+	echo ${mac_address} | grep -q -P "^([0-9a-f]{2}:){5}[0-9a-f]{2}$"
+	if [ $? -ne 0 ]; then
+		ocf_log error "MAC address '${mac_address}' is not valid."
+		return $OCF_ERR_GENERIC
+	fi
+
+	if ! $OCF_RESKEY_openstackcli port unset \
+		--allowed-address \
+		ip-address=$OCF_RESKEY_ip,mac-address=${mac_address} \
+		$node_port_id; then
+		return $OCF_ERR_GENERIC
+	fi
+
+	osvip_monitor
+	if [ $? != $OCF_NOT_RUNNING ]; then
+		ocf_log error "Couldn't unset IP address $OCF_RESKEY_ip."
+		return $OCF_ERR_GENERIC
+	fi
+
+	ocf_log info "Successfully brought unset $OCF_RESKEY_ip"
+	return $OCF_SUCCESS
+}
+
+osvip_start() {
+	node_port_id=$(osvip_port_id)
+
+	osvip_monitor
+	if [ $? = $OCF_SUCCESS ]; then
+		ocf_log info "$OCF_RESKEY_ip already started"
+		return $OCF_SUCCESS
+	fi
+
+	ocf_log info "Moving IP address $OCF_RESKEY_ip to port ID $node_port_id"
+
+	$OCF_RESKEY_openstackcli port set \
+		--allowed-address ip-address=$OCF_RESKEY_ip \
+		$node_port_id
+	if [ $? != $OCF_SUCCESS ]; then
+		ocf_log error "$OCF_RESKEY_ip Cannot be set to port $node_port_id"
+		return $OCF_ERR_GENERIC
+	fi
+
+	osvip_monitor
+	if [ $? != $OCF_SUCCESS ]; then
+		ocf_log error "$OCF_RESKEY_ip Cannot be set to port $node_port_id"
+		return $OCF_ERR_GENERIC
+	fi
+
+	return $OCF_SUCCESS
+}
+
+###############################################################################
+#
+# MAIN
+#
+###############################################################################
+
+case $__OCF_ACTION in
+	meta-data)
+		metadata
+		exit $OCF_SUCCESS
+		;;
+	usage|help)
+		echo $USAGE
+		exit $OCF_SUCCESS
+		;;
+esac
+
+if ! ocf_is_root; then
+	ocf_log err "You must be root for $__OCF_ACTION operation."
+	exit $OCF_ERR_PERM
+fi
+
+osvip_validate
```

Comment:
You'll have to use exit in the validate function, or the result will be ignored when action isnt validate-all.

Reply:
Fixed with commit 61aebd2c482b0052b2865fba01acc11bc0fc5fab:
```
-osvip_validate
-
 case $__OCF_ACTION in
 	start)
-		osvip_start;;
+		osvip_validate
+		osvip_start
+		;;
 	stop)
 		osvip_stop;;
 	monitor)
 		osvip_monitor;;
 	validate-all)
-		exit $?;;
+		osvip_validate;;
```

Reply:
I actually changed this again (once I realize I broke the agent with the change).  Now, osvip_validate is still removed from the main and has been added to the beginning of the monitor function.  The reason for this is that start|stop|monitor all require the openstack CLI and the resource file that the validation checks.  The sourcing of the resource file also happens in the validation function, so by adding validation to monitoring and all the appropriate functions call monitoring, the proper checks and environment variables are now set.